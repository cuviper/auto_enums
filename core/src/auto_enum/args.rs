use proc_macro2::{token_stream::IntoIter, Delimiter, Ident, TokenStream, TokenTree};
use syn::{Path, Result};

macro_rules! error {
    ($span:expr, $msg:expr) => {
         Err(syn::Error::new($span.span(), $msg))
    };
    ($span:expr, $($tt:tt)*) => {
        error!($span, format!($($tt)*))
    };
}

pub(super) fn parse_args(args: TokenStream) -> Result<(Vec<Path>, Option<String>)> {
    const ERR: &str = "expected one of `,`, `::`, or identifier, found ";

    let mut iter = args.into_iter();
    let mut args = Vec::new();
    let mut marker = None;
    while let Some(tt) = iter.next() {
        match tt {
            TokenTree::Ident(i) => {
                if i == "marker" {
                    marker_opt(i, &mut iter, &mut args, &mut marker)?;
                } else {
                    args.push(parse_path(i.into(), None, &mut iter)?);
                }
            }
            TokenTree::Punct(p) => match p.as_char() {
                ',' => {}
                ':' => args.push(parse_path(p.into(), None, &mut iter)?),
                _ => return error!(p, "{}`{}`", ERR, p),
            },
            _ => return error!(tt, "{}`{}`", ERR, tt),
        }
    }

    Ok((args, marker))
}

fn parse_path(tt1: TokenTree, tt2: Option<TokenTree>, iter: &mut IntoIter) -> Result<Path> {
    let mut tokens = TokenStream::new();
    tokens.extend(Some(tt1));
    tokens.extend(tt2);
    let iter =
        iter.take_while(|tt| if let TokenTree::Punct(p) = tt { p.as_char() != ',' } else { true });
    tokens.extend(iter);

    syn::parse2(tokens)
}

fn marker_opt(
    ident: Ident,
    iter: &mut IntoIter,
    args: &mut Vec<Path>,
    marker: &mut Option<String>,
) -> Result<()> {
    match iter.next() {
        Some(TokenTree::Group(ref g)) if g.delimiter() != Delimiter::Parenthesis => {
            return error!(g, "invalid delimiter")
        }
        Some(TokenTree::Group(g)) => {
            let mut tokens = g.stream().into_iter();
            let i = match tokens.next() {
                Some(TokenTree::Ident(ref i)) if marker.is_some() => {
                    return error!(i, "multiple `marker` option")
                }
                Some(TokenTree::Ident(i)) => i,
                Some(tt) => return error!(tt, "expected an identifier, found `{}`", tt),
                None => return error!(g, "empty `marker` option"),
            };
            *marker = Some(i.to_string());
            match tokens.next() {
                None => {}
                Some(TokenTree::Punct(ref p)) if p.as_char() == ',' => {
                    if let Some(tt) = tokens.next() {
                        // TODO: https://docs.rs/proc-macro2/0.4/proc_macro2/struct.Span.html#method.join
                        // `i.span().join(tt.span()).unwrap_or_else(|| tt.span())`
                        return error!(tt, "multiple identifier in `marker` option");
                    }
                }
                // TODO: https://docs.rs/proc-macro2/0.4/proc_macro2/struct.Span.html#method.join
                Some(tt) => return error!(tt, "multiple identifier in `marker` option"),
            }
        }
        Some(TokenTree::Punct(ref p)) if p.as_char() == '=' => {
            match iter.next() {
                Some(TokenTree::Ident(ref i)) if marker.is_some() => {
                    return error!(i, "multiple `marker` option")
                }
                Some(TokenTree::Ident(i)) => *marker = Some(i.to_string()),
                Some(tt) => return error!(tt, "expected an identifier, found `{}`", tt),
                None => return error!(p, "empty `marker` option"),
            }
            match iter.next() {
                None => {}
                Some(TokenTree::Punct(ref p)) if p.as_char() == ',' => {}
                Some(tt) => return error!(tt, "expected `,`, found `{}`", tt),
            }
        }
        tt => args.push(parse_path(ident.into(), tt, iter)?),
    }

    Ok(())
}
