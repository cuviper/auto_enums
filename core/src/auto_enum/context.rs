use std::cell::RefCell;

use proc_macro2::{Ident, TokenStream};
use quote::ToTokens;
use smallvec::{smallvec, SmallVec};
use syn::{token, Attribute, Expr, ExprCall, ExprPath, ItemEnum, Macro, Result};

use crate::utils::{ident, path, Stack};

use super::{
    visitor::{Dummy, FindTry, Visitor},
    *,
};

// =============================================================================
// Context

/// Config for related to `visotor::Visotor` type.
#[derive(Clone, Copy, PartialEq, Eq)]
pub(super) enum VisitMode {
    Default,
    Return,
    Try,
}

/// Config for related to `expr::VisitLast` trait.
#[derive(Clone, Copy, PartialEq, Eq)]
pub(super) enum VisitLastMode {
    Default,
    /*
    local `let .. = || {};`
    or
    expr `|| {}`
    not
    item_fn `fn _() -> Fn*() { || {} }`
    */
    Closure,
    /// `Stmt::Semi(_, _)` or `never` option - never visit last expr
    Never,
}

pub(super) struct Context {
    /// Span passed to `syn::Error::new_spanned`.
    span: Option<TokenStream>,
    pub(super) args: Stack<Arg>,
    builder: Builder,
    pub(super) marker: Marker,
    // pub(super) depth: usize,
    root: bool,
    pub(super) attr: bool,
    mode: VisitMode,
    visit_last: VisitLastMode,
    /// This in `true` if error occurred in visiting.
    pub(super) error: bool,
}

impl Context {
    fn new<T: ToTokens>(
        span: T,
        args: Stack<Arg>,
        marker: Option<String>,
        never: bool,
        root: bool,
    ) -> Self {
        Self {
            span: Some(span.into_token_stream()),
            args,
            builder: Builder::new(),
            marker: Marker::new(marker),
            // depth: 0,
            root,
            attr: false,
            mode: VisitMode::Default,
            visit_last: if never { VisitLastMode::Never } else { VisitLastMode::Default },
            error: false,
        }
    }

    pub(super) fn root<T: ToTokens>(
        span: T,
        (args, marker, never): (Stack<Arg>, Option<String>, bool),
    ) -> Self {
        Self::new(span, args, marker, never, true)
    }

    pub(super) fn child<T: ToTokens>(
        span: T,
        (args, marker, never): (Stack<Arg>, Option<String>, bool),
    ) -> Self {
        Self::new(span, args, marker, never, false)
    }

    // If this is called more than once, it is a bug.
    pub(super) fn span(&mut self) -> TokenStream {
        self.span.take().unwrap_or_else(|| unreachable!())
    }

    pub(super) const fn mode(&self) -> VisitMode {
        self.mode
    }

    pub(super) fn visit_mode(&mut self, mode: VisitMode) {
        self.mode = mode;
    }

    pub(super) fn visit_last_mode(&mut self, visit_last: VisitLastMode) {
        self.visit_last = visit_last;
    }

    pub(super) fn visit_last(&self) -> bool {
        self.visit_last != VisitLastMode::Never && self.mode != VisitMode::Try
    }

    pub(super) fn visitor<F: FnOnce(&mut Visitor<'_>)>(&mut self, f: F) {
        f(&mut Visitor::new(self));
    }

    pub(super) fn dummy<F: FnOnce(&mut Dummy<'_>)>(&mut self, f: F) {
        f(&mut Dummy::new(self));
    }

    pub(super) fn find_try<F: FnOnce(&mut FindTry<'_>)>(&mut self, f: F) {
        let mut find = FindTry::new(self);
        f(&mut find);
        if find.has {
            self.mode = VisitMode::Try;
        }
    }

    pub(super) fn next_expr(&mut self, expr: Expr) -> Expr {
        self.next_expr_with_attrs(Vec::new(), expr)
    }

    pub(super) fn next_expr_with_attrs(&mut self, attrs: Vec<Attribute>, expr: Expr) -> Expr {
        self.builder.next_expr(attrs, expr)
    }

    pub(super) fn marker_macro(&self, Macro { path, .. }: &Macro) -> bool {
        match &self.marker.ident {
            None => path.is_ident(DEFAULT_MARKER),
            Some(marker) => path.is_ident(marker) || (!self.root && path.is_ident(DEFAULT_MARKER)),
        }
    }

    /* FIXME: This may not be necessary.
    pub(super) fn assigned_enum(&self, ExprCall { args, func, .. }: &ExprCall) -> bool {
        args.len() == 1
            && match &**func {
                Expr::Path(ExprPath {
                    path, qself: None, ..
                }) => {
                    path.leading_colon.is_none()
                        && path.segments.len() == 2
                        && path.segments[0].arguments.is_empty()
                        && path.segments[1].arguments.is_empty()
                        && path.segments[0].ident == self.builder.ident
                }
                _ => false,
            }
    }
    */

    fn buildable(&mut self) -> Result<bool> {
        fn err(cx: &mut Context, len: usize) -> Result<bool> {
            let (msg1, msg2) = match cx.visit_last {
                VisitLastMode::Default | VisitLastMode::Closure => {
                    ("branches or marker macros in total", "branch or marker macro")
                }
                VisitLastMode::Never => ("marker macros", "marker macro"),
            };

            Err(err!(cx.span(),
                "the `#[auto_enum]` attribute is required two or more {}, there is {} {} in this statement",
                msg1,
                if len == 0 { "no" } else { "only one" },
                msg2
            ))?
        }

        if self.error {
            // As we know that an error will occur, it does not matter if there are not enough variants.
            Ok(true)
        } else {
            match self.builder.variants.len() {
                1 => err(self, 1),
                0 if !self.attr => err(self, 0),
                0 => Ok(false),
                _ => Ok(true),
            }
        }
    }

    pub(super) fn build<F: FnOnce(ItemEnum)>(&mut self, f: F) -> Result<()> {
        self.buildable().map(|buildable| {
            if buildable {
                f(self.builder.build(&self.args))
            }
        })
    }

    #[cfg(feature = "type_analysis")]
    pub(super) fn impl_traits(&mut self, ty: &mut Type) {
        collect_impl_traits(&mut self.args, ty);
    }
}

// =============================================================================
// Expression level marker

pub(super) const DEFAULT_MARKER: &str = "marker";

pub(super) struct Marker {
    ident: Option<String>,
}

impl Marker {
    const fn new(ident: Option<String>) -> Self {
        Self { ident }
    }

    pub(super) fn is_unique(&self) -> bool {
        self.ident.is_some()
    }

    pub(super) fn ident(&self) -> &str {
        self.ident.as_ref().map_or(DEFAULT_MARKER, |s| s)
    }
}

// =============================================================================
// Enum builder

struct Builder {
    ident: String,
    variants: Stack<String>,
}

impl Builder {
    fn new() -> Self {
        Self {
            ident: format!("___Enum{}", RNG.with(|rng| rng.borrow_mut().next())),
            variants: Stack::new(),
        }
    }

    fn iter(&self) -> impl Iterator<Item = Ident> + '_ {
        self.variants.iter().map(ident)
    }

    fn next_expr(&mut self, attrs: Vec<Attribute>, expr: Expr) -> Expr {
        let variant = format!("___Variant{}", self.variants.len());

        let segments: SmallVec<[_; 2]> =
            smallvec![ident(&self.ident).into(), ident(&variant).into()];

        self.variants.push(variant);

        Expr::Call(ExprCall {
            attrs,
            func: Box::new(Expr::Path(ExprPath {
                attrs: Vec::new(),
                qself: None,
                path: path(segments),
            })),
            paren_token: token::Paren::default(),
            args: Some(expr).into_iter().collect(),
        })
    }

    fn build(&self, args: &[Arg]) -> ItemEnum {
        let ident = ident(&self.ident);
        let ty_generics = self.iter();
        let variants = self.iter();
        let fields = self.iter();

        syn::parse_quote! {
            #[::auto_enums::enum_derive(#(#args),*)]
            enum #ident<#(#ty_generics),*> {
                #(#variants(#fields),)*
            }
        }
    }
}

// =============================================================================
// RNG

thread_local! {
    static RNG: RefCell<XorShift64Star> = RefCell::new(XorShift64Star::new());
}

// https://github.com/rayon-rs/rayon/blob/rayon-core-v1.4.1/rayon-core/src/registry.rs#L712-L750

use std::{
    cell::Cell,
    collections::hash_map::DefaultHasher,
    hash::Hasher,
    sync::atomic::{AtomicUsize, Ordering},
};

/// [xorshift*] is a fast pseudorandom number generator which will
/// even tolerate weak seeding, as long as it's not zero.
///
/// [xorshift*]: https://en.wikipedia.org/wiki/Xorshift#xorshift*
struct XorShift64Star {
    state: Cell<u64>,
}

impl XorShift64Star {
    fn new() -> Self {
        // Any non-zero seed will do -- this uses the hash of a global counter.
        let mut seed = 0;
        while seed == 0 {
            static COUNTER: AtomicUsize = AtomicUsize::new(0);
            let mut hasher = DefaultHasher::new();
            hasher.write_usize(COUNTER.fetch_add(1, Ordering::Relaxed));
            seed = hasher.finish();
        }

        Self { state: Cell::new(seed) }
    }

    fn next(&self) -> u64 {
        let mut x = self.state.get();
        debug_assert_ne!(x, 0);
        x ^= x >> 12;
        x ^= x << 25;
        x ^= x >> 27;
        self.state.set(x);
        x.wrapping_mul(0x2545_f491_4f6c_dd1d)
    }
}
