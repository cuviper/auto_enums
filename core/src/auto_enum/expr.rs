use syn::{
    visit_mut::{self, VisitMut},
    *,
};

use crate::utils::{expr_block, replace_block, replace_expr, Attrs};

use super::{Context, NAME, NESTED, NEVER};

/// Visits last expression.
///
/// Note that do not use this after `cx.visitor()`.
pub(super) fn child_expr(expr: &mut Expr, cx: &mut Context) -> Result<()> {
    if !cx.visit_last() || is_unreachable(cx, expr) {
        return Ok(());
    }

    match expr {
        Expr::Block(ExprBlock { block, .. }) | Expr::Unsafe(ExprUnsafe { block, .. }) => {
            if let Some(Stmt::Expr(expr)) = block.stmts.last_mut() {
                return child_expr(expr, cx);
            }
        }
        _ => {}
    }

    match expr {
        Expr::Match(expr) => visit_last_expr_match(cx, expr),
        Expr::If(expr) => visit_last_expr_if(cx, expr),
        Expr::Loop(expr) => visit_last_expr_loop(cx, expr),

        // Search recursively
        Expr::MethodCall(ExprMethodCall { receiver: expr, .. })
        | Expr::Paren(ExprParen { expr, .. })
        | Expr::Type(ExprType { expr, .. }) => child_expr(expr, cx),

        _ => Ok(()),
    }
}

pub(super) fn is_unreachable(cx: &Context, expr: &Expr) -> bool {
    const UNREACHABLE_MACROS: &[&str] = &["unreachable", "panic"];

    if expr.any_empty_attr(NEVER) || expr.any_attr(NAME) {
        return true;
    }

    match expr {
        Expr::Block(ExprBlock { block, .. }) | Expr::Unsafe(ExprUnsafe { block, .. }) => {
            is_unreachable_stmt(cx, block.stmts.last())
        }

        Expr::Break(_) | Expr::Continue(_) | Expr::Return(_) => true,

        // `unreachable!`, `panic!` or an expression level marker (`marker!` macro).
        Expr::Macro(ExprMacro { mac, .. }) => {
            UNREACHABLE_MACROS.iter().any(|i| mac.path.is_ident(i)) || cx.is_marker_macro(mac)
        }

        Expr::Match(ExprMatch { arms, .. }) => {
            arms.iter().all(|arm| arm.any_empty_attr(NEVER) || is_unreachable(cx, &*arm.body))
        }

        // `Err(expr)?` or `None?`.
        Expr::Try(ExprTry { expr, .. }) => match &**expr {
            Expr::Path(ExprPath { path, qself: None, .. }) => path.is_ident("None"),
            Expr::Call(ExprCall { args, func, .. }) if args.len() == 1 => match &**func {
                Expr::Path(ExprPath { path, qself: None, .. }) => path.is_ident("Err"),
                _ => false,
            },
            _ => false,
        },

        // Search recursively
        Expr::MethodCall(ExprMethodCall { receiver: expr, .. })
        | Expr::Paren(ExprParen { expr, .. })
        | Expr::Type(ExprType { expr, .. }) => is_unreachable(cx, expr),

        _ => false,
    }
}

fn is_unreachable_stmt(cx: &Context, stmt: Option<&Stmt>) -> bool {
    match stmt {
        Some(Stmt::Expr(expr)) | Some(Stmt::Semi(expr, _)) => is_unreachable(cx, expr),
        Some(Stmt::Local(local)) => {
            local.init.as_ref().map_or(false, |(_, expr)| is_unreachable(cx, expr))
        }
        Some(Stmt::Item(_)) => true,
        None => false,
    }
}

fn visit_last_expr_match(cx: &mut Context, expr: &mut ExprMatch) -> Result<()> {
    fn skip(arm: &mut Arm, cx: &mut Context) -> bool {
        arm.any_empty_attr(NEVER)
            || arm.any_empty_attr(NESTED)
            || is_unreachable(cx, &*arm.body)
            || cx.find_nested(arm)
    }

    expr.arms.iter_mut().try_for_each(|arm| {
        if !skip(arm, cx) {
            arm.comma = Some(token::Comma::default());
            replace_expr(&mut *arm.body, |x| cx.next_expr(x));
        }
        Ok(())
    })
}

fn visit_last_expr_if(cx: &mut Context, expr: &mut ExprIf) -> Result<()> {
    fn skip(block: &mut Block, cx: &mut Context) -> bool {
        match block.stmts.last_mut() {
            Some(Stmt::Expr(expr)) => {
                expr.any_empty_attr(NESTED) || is_unreachable(cx, expr) || cx.find_nested(block)
            }
            _ => is_unreachable_stmt(cx, block.stmts.last()),
        }
    }

    if !skip(&mut expr.then_branch, cx) {
        replace_block(&mut expr.then_branch, |b| cx.next_expr(expr_block(b)));
    }

    match expr.else_branch.as_mut().map(|(_, expr)| &mut **expr) {
        Some(Expr::Block(expr)) => {
            if !skip(&mut expr.block, cx) {
                replace_block(&mut expr.block, |b| cx.next_expr(expr_block(b)));
            }
            Ok(())
        }
        Some(Expr::If(expr)) => visit_last_expr_if(cx, expr),

        // TODO: https://docs.rs/proc-macro2/0.4/proc_macro2/struct.Span.html#method.join
        // `expr.span().join(expr.then_branch.span()).unwrap_or_else(|| expr.span())``
        None => Err(error!(expr.if_token, "`if` expression missing an else clause")),

        Some(_) => unreachable!("wrong_if"),
    }
}

fn visit_last_expr_loop(cx: &mut Context, expr: &mut ExprLoop) -> Result<()> {
    LoopVisitor::new(expr.label.as_ref(), cx).visit_block_mut(&mut expr.body);
    Ok(())
}

// =================================================================================================
// LoopVisitor

struct LoopVisitor<'a> {
    cx: &'a mut Context,
    label: Option<&'a Label>,
    nested: bool,
}

impl<'a> LoopVisitor<'a> {
    fn new(label: Option<&'a Label>, cx: &'a mut Context) -> Self {
        Self { cx, label, nested: false }
    }

    fn compare_labels(&self, other: Option<&Lifetime>) -> bool {
        match (self.label, other) {
            (None, None) => true,
            (Some(this), Some(other)) => this.name.ident == other.ident,
            _ => false,
        }
    }
}

impl VisitMut for LoopVisitor<'_> {
    fn visit_expr_mut(&mut self, node: &mut Expr) {
        if node.any_empty_attr(NEVER) {
            return;
        }

        let tmp = self.nested;
        match node {
            // Stop at closure / async block bounds
            Expr::Closure(_) | Expr::Async(_) => return,

            // Other loop bounds
            Expr::Loop(_) | Expr::ForLoop(_) | Expr::While(_) => {
                if self.label.is_none() {
                    return;
                }
                self.nested = true;
            }

            // Desugar `break <expr>` into `break Enum::VariantN(<expr>)`.
            Expr::Break(ExprBreak { label, expr, .. }) => {
                if !self.nested && label.is_none() || self.compare_labels(label.as_ref()) {
                    self.cx.replace_boxed_expr(expr);
                }
            }

            _ => {}
        }

        visit_mut::visit_expr_mut(self, node);
        self.nested = tmp;
    }

    fn visit_item_mut(&mut self, _: &mut Item) {
        // Do not recurse into nested items.
    }
}
