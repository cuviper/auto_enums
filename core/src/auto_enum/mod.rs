use proc_macro2::TokenStream;
use quote::ToTokens;
use syn::*;

use crate::utils::*;

mod context;
mod expr;
#[cfg(feature = "type_analysis")]
mod type_analysis;
mod visitor;

use self::{
    context::{Context, VisitLastMode, VisitMode, DEFAULT_MARKER},
    expr::child_expr,
};

/// The attribute name.
const NAME: &str = "auto_enum";

/// The annotation for recursively parsing.
const NESTED: &str = "nested";
/// The annotation for skipping branch.
const NEVER: &str = "never";

/// The old annotation replaced by `#[nested]`.
const NESTED_OLD: &str = "rec";

pub(crate) fn attribute(args: TokenStream, input: TokenStream) -> TokenStream {
    expand(args, input)
}

fn expand(args: TokenStream, input: TokenStream) -> TokenStream {
    let mut cx = match Context::root(input.clone(), args) {
        Err(e) => return e.to_compile_error(),
        Ok(cx) => cx,
    };

    let res = match syn::parse2::<Stmt>(input.clone()) {
        Ok(mut stmt) => expand_parent_stmt(&mut stmt, &mut cx).map(|()| stmt.into_token_stream()),
        Err(e) => syn::parse2::<Expr>(input)
            .map_err(|_e| {
                cx.diagnostic.error(e);
                error!(cx.span, "may only be used on expression, statement, or function")
            })
            .and_then(|mut expr| {
                expand_parent_expr(&mut expr, &mut cx).map(|()| expr.into_token_stream())
            }),
    };

    match res {
        Err(e) => cx.diagnostic.error(e),
        Ok(_) if cx.failed() => {}
        Ok(tokens) => return tokens,
    }
    cx.diagnostic.to_compile_error().unwrap()
}

fn visit_expr(expr: &mut Expr, cx: &mut Context) -> Result<()> {
    let expr = match expr {
        Expr::Closure(ExprClosure { body, .. }) if cx.visit_last() => {
            cx.visit_mode = VisitMode::Return;
            cx.visit_last_mode = VisitLastMode::Closure;
            cx.find_try(&mut **body);
            &mut **body
        }
        _ => expr,
    };

    child_expr(expr, cx).map(|()| cx.visitor(expr))
}

fn build_expr(expr: &mut Expr, item: ItemEnum) {
    replace_expr(expr, |expr| expr_block(block(vec![Stmt::Item(item.into()), Stmt::Expr(expr)])));
}

// =================================================================================================
// Expand statement or expression in which `#[auto_enum]` was directly used.

fn expand_parent_stmt(stmt: &mut Stmt, cx: &mut Context) -> Result<()> {
    if let Stmt::Semi(..) = &stmt {
        cx.visit_last_mode = VisitLastMode::Never;
    }

    match stmt {
        Stmt::Expr(expr) | Stmt::Semi(expr, _) => expand_parent_expr(expr, cx),
        Stmt::Local(local) => expand_parent_local(local, cx),
        Stmt::Item(Item::Fn(item)) => expand_parent_item_fn(item, cx),
        Stmt::Item(item) => {
            Err(error!(item, "may only be used on expression, statement, or function"))
        }
    }
}

fn expand_parent_expr(expr: &mut Expr, cx: &mut Context) -> Result<()> {
    if cx.is_dummy() {
        cx.dummy(expr);
        return Ok(());
    }

    visit_expr(expr, cx)?;

    cx.build(|item| build_expr(expr, item))
}

fn expand_parent_local(local: &mut Local, cx: &mut Context) -> Result<()> {
    #[cfg(feature = "type_analysis")]
    {
        if let Pat::Type(PatType { ty, .. }) = &mut local.pat {
            cx.collect_trait(&mut *ty);
        }
    }

    if cx.is_dummy() {
        cx.dummy(local);
        return Ok(());
    }

    let expr = if let Some((_, expr)) = &mut local.init {
        &mut **expr
    } else {
        return Err(error!(
            local,
            "the `#[auto_enum]` attribute is not supported uninitialized let statement"
        ));
    };

    visit_expr(expr, cx)?;

    cx.build(|item| build_expr(expr, item))
}

fn expand_parent_item_fn(item: &mut ItemFn, cx: &mut Context) -> Result<()> {
    let ItemFn { sig, block, .. } = item;
    if let ReturnType::Type(_, ty) = &mut sig.output {
        match &**ty {
            // `return`
            Type::ImplTrait(_) if cx.visit_last() => cx.visit_mode = VisitMode::Return,

            // `?` operator
            Type::Path(TypePath { qself: None, path }) if cx.visit_last() => {
                let PathSegment { arguments, ident } = &path.segments[path.segments.len() - 1];
                match arguments {
                    // `Result<T, impl Trait>`
                    PathArguments::AngleBracketed(AngleBracketedGenericArguments {
                        colon2_token: None,
                        args,
                        ..
                    }) if args.len() == 2 && ident == "Result" => {
                        if let (
                            GenericArgument::Type(_),
                            GenericArgument::Type(Type::ImplTrait(_)),
                        ) = (&args[0], &args[1])
                        {
                            cx.find_try(&mut **block);
                        }
                    }
                    _ => {}
                }
            }

            _ => {}
        }

        #[cfg(feature = "type_analysis")]
        cx.collect_trait(&mut *ty);
    }

    if cx.is_dummy() {
        cx.dummy(item);
        return Ok(());
    }

    match item.block.stmts.last_mut() {
        Some(Stmt::Expr(expr)) => child_expr(expr, cx)?,
        Some(_) => {}
        None => {
            return Err(error!(
                item.block,
                "the `#[auto_enum]` attribute is not supported empty functions"
            ));
        }
    }

    cx.visitor(item);

    cx.build(|i| item.block.stmts.insert(0, Stmt::Item(i.into())))
}
