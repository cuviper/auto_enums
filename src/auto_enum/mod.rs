mod context;
mod expr;
#[cfg(feature = "type_analysis")]
mod type_analysis;
mod visitor;

use proc_macro2::TokenStream;
use quote::ToTokens;
#[cfg(feature = "type_analysis")]
use syn::Pat;
use syn::{
    AngleBracketedGenericArguments, Expr, ExprClosure, GenericArgument, Item, ItemEnum, ItemFn,
    Local, PathArguments, Result, ReturnType, Stmt, Type, TypePath,
};

use self::{
    context::{Context, VisitLastMode, VisitMode, DEFAULT_MARKER},
    expr::child_expr,
};
use crate::utils::{block, expr_block, replace_expr};

/// The attribute name.
const NAME: &str = "auto_enum";
/// The annotation for recursively parsing.
const NESTED: &str = "nested";
/// The annotation for skipping branch.
const NEVER: &str = "never";

pub(crate) fn attribute(args: TokenStream, input: TokenStream) -> TokenStream {
    let mut cx = match Context::root(input.clone(), args) {
        Err(e) => return e.to_compile_error(),
        Ok(cx) => cx,
    };

    let res = match syn::parse2::<Stmt>(input.clone()) {
        Ok(mut stmt) => expand_parent_stmt(&mut cx, &mut stmt).map(|()| stmt.into_token_stream()),
        Err(e) => syn::parse2::<Expr>(input)
            .map_err(|_e| {
                cx.error(e);
                format_err!(cx.span, "may only be used on expression, statement, or function")
            })
            .and_then(|mut expr| {
                expand_parent_expr(&mut cx, &mut expr, false).map(|()| expr.into_token_stream())
            }),
    };

    match res {
        Err(e) => cx.error(e),
        Ok(_) if cx.has_error() => {}
        Ok(tokens) => return tokens,
    }
    cx.compile_error().unwrap()
}

fn expand_expr(cx: &mut Context, expr: &mut Expr) -> Result<()> {
    let expr = match expr {
        Expr::Closure(ExprClosure { body, .. }) if cx.visit_last() => {
            let count = visitor::visit_fn(cx, &mut **body);
            if count.try_ >= 2 {
                cx.visit_mode = VisitMode::Try;
            } else {
                cx.visit_mode = VisitMode::Return(count.return_);
            }
            &mut **body
        }
        _ => expr,
    };

    child_expr(cx, expr)?;

    #[cfg(feature = "type_analysis")]
    {
        if let VisitMode::Return(count) = cx.visit_mode {
            if cx.args.is_empty() && cx.variant_is_empty() && count < 2 {
                cx.dummy(expr);
                return Ok(());
            }
        }
    }

    cx.visitor(expr);

    Ok(())
}

fn build_expr(expr: &mut Expr, item: ItemEnum) {
    replace_expr(expr, |expr| expr_block(block(vec![Stmt::Item(item.into()), Stmt::Expr(expr)])));
}

// =================================================================================================
// Expand statement or expression in which `#[auto_enum]` was directly used.

fn expand_parent_stmt(cx: &mut Context, stmt: &mut Stmt) -> Result<()> {
    match stmt {
        Stmt::Expr(expr) => expand_parent_expr(cx, expr, false),
        Stmt::Semi(expr, _) => expand_parent_expr(cx, expr, true),
        Stmt::Local(local) => expand_parent_local(cx, local),
        Stmt::Item(Item::Fn(item)) => expand_parent_item_fn(cx, item),
        Stmt::Item(item) => {
            bail!(item, "may only be used on expression, statement, or function");
        }
    }
}

fn expand_parent_expr(cx: &mut Context, expr: &mut Expr, has_semi: bool) -> Result<()> {
    if has_semi {
        cx.visit_last_mode = VisitLastMode::Never;
    }

    if cx.is_dummy() {
        cx.dummy(expr);
        return Ok(());
    }

    expand_expr(cx, expr)?;

    cx.build(|item| build_expr(expr, item))
}

fn expand_parent_local(cx: &mut Context, local: &mut Local) -> Result<()> {
    #[cfg(feature = "type_analysis")]
    {
        if let Pat::Type(pat) = &mut local.pat {
            if cx.collect_impl_trait(&mut pat.ty) {
                local.pat = (*pat.pat).clone()
            }
        }
    }

    if cx.is_dummy() {
        cx.dummy(local);
        return Ok(());
    }

    let expr = if let Some((_, expr)) = &mut local.init {
        &mut **expr
    } else {
        bail!(local, "the `#[auto_enum]` attribute is not supported uninitialized let statement");
    };

    expand_expr(cx, expr)?;

    cx.build(|item| build_expr(expr, item))
}

fn expand_parent_item_fn(cx: &mut Context, item: &mut ItemFn) -> Result<()> {
    let ItemFn { sig, block, .. } = item;
    if let ReturnType::Type(_, ty) = &mut sig.output {
        match &**ty {
            // `return`
            Type::ImplTrait(_) if cx.visit_last_mode != VisitLastMode::Never => {
                let count = visitor::visit_fn(cx, &mut **block);
                cx.visit_mode = VisitMode::Return(count.return_);
            }

            // `?` operator
            Type::Path(TypePath { qself: None, path })
                if cx.visit_last_mode != VisitLastMode::Never =>
            {
                let ty = path.segments.last().unwrap();
                match &ty.arguments {
                    // `Result<T, impl Trait>`
                    PathArguments::AngleBracketed(AngleBracketedGenericArguments {
                        colon2_token: None,
                        args,
                        ..
                    }) if args.len() == 2 && ty.ident == "Result" => {
                        if let (
                            GenericArgument::Type(_),
                            GenericArgument::Type(Type::ImplTrait(_)),
                        ) = (&args[0], &args[1])
                        {
                            let count = visitor::visit_fn(cx, &mut **block);
                            if count.try_ >= 2 {
                                cx.visit_mode = VisitMode::Try;
                            }
                        }
                    }
                    _ => {}
                }
            }

            _ => {}
        }

        #[cfg(feature = "type_analysis")]
        cx.collect_impl_trait(&mut *ty);
    }

    if cx.is_dummy() {
        cx.dummy(item);
        return Ok(());
    }

    match item.block.stmts.last_mut() {
        Some(Stmt::Expr(expr)) => child_expr(cx, expr)?,
        Some(_) => {}
        None => {
            bail!(item.block, "the `#[auto_enum]` attribute is not supported empty functions");
        }
    }

    #[cfg(feature = "type_analysis")]
    {
        if let VisitMode::Return(count) = cx.visit_mode {
            if cx.args.is_empty() && cx.variant_is_empty() && count < 2 {
                cx.dummy(item);
                return Ok(());
            }
        }
    }

    cx.visitor(item);

    cx.build(|i| item.block.stmts.insert(0, Stmt::Item(i.into())))
}
