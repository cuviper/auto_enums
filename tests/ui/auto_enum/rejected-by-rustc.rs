// compile-fail

#![feature(try_trait)]

use auto_enums::auto_enum;

#[auto_enum(Iterator)]
fn wrong_if(x: usize) -> impl Iterator<Item = i32> {
    if x == 0 {
        1..8
    } else return {
        //~^ ERROR expected `{`, found keyword `return`
        2..=10
    }
}

fn main() {}
