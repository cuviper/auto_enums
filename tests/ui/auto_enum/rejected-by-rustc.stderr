error: expected `{`, found keyword `return`
 --> $DIR/rejected-by-rustc.rs:7:12
  |
7 |     } else return {
  |            ^^^^^^ expected `{`
help: try placing this code inside a block
  |
7 |     } else { return {
8 |         //~^ ERROR expected `{`, found keyword `return`
9 |         2..=10
10|     } }
  |

error: attributes are not yet allowed on `if` expressions
  --> $DIR/rejected-by-rustc.rs:18:9
   |
18 |         #[nested]
   |         ^^^^^^^^^
